File: src/main/java/com/everchanging/hierarchy/converter/TopEmployeeToHierarchyMapConverter.java
package com.everchanging.hierarchy.converter;

import com.everchanging.hierarchy.model.Employee;
import lombok.extern.slf4j.Slf4j;
import org.springframework.core.convert.converter.Converter;
import org.springframework.stereotype.Component;

import javax.validation.constraints.NotNull;
import java.util.Map;
import java.util.Set;

import static java.util.stream.Collectors.toMap;

@Slf4j
@Component
public class TopEmployeeToHierarchyMapConverter implements Converter<Employee, Map<String, Object>> {
    private static final Object EMPTY_OBJECT = new Object();

    @Override
    public Map<String, Object> convert(@NotNull Employee topLevelEmployee) {
        log.debug("Convert from top topLevelEmployee {} to hierarchy map", topLevelEmployee);
        if (topLevelEmployee.getSupervisor() != null) {
            log.error("Received a non-top topLevelEmployee to convert into hierarchy map. {}", topLevelEmployee);
            throw new IllegalStateException();
        }

        return Map.of(topLevelEmployee.getName(), convertList(topLevelEmployee.getSubordinates()));
    }

    public Object convertList(Set<Employee> employees) {
        if (employees.isEmpty()) {
            return EMPTY_OBJECT;
        }
        return employees.stream()
                .collect(toMap(Employee::getName, employee -> convertList(employee.getSubordinates())));
    }
}

File: src/main/kotlin/com/everchanging/hierarchy/converter/TopEmployeeToHierarchyMapConverter.kt
package com.everchanging.hierarchy.converter

import com.everchanging.hierarchy.model.Employee
import org.slf4j.LoggerFactory
import org.springframework.core.convert.converter.Converter
import org.springframework.stereotype.Component

@Component
class TopEmployeeToHierarchyMapConverter : Converter<Employee, Map<String, Any>> {
    companion object {
        private val log = LoggerFactory.getLogger(TopEmployeeToHierarchyMapConverter::class.java)
    }

    private val emptyObject: Any = Object()

    override fun convert(topLevelEmployee: Employee): Map<String, Any> {
        log.debug("Convert from top topLevelEmployee $topLevelEmployee to hierarchy map")
        if (topLevelEmployee.supervisor != null) {
            log.error("Received a non-top topLevelEmployee to convert into hierarchy map. $topLevelEmployee")
            throw IllegalStateException()
        }

        return mapOf(topLevelEmployee.name to convertList(topLevelEmployee.subordinates))
    }

    private fun convertList(employees: Set<Employee>): Any {
        if (employees.isEmpty()) {
            return emptyObject
        }

        return employees.map { it.name to convertList(it.subordinates) }.toMap()
    }
}